src/inmanta/agent/agent.py

```diff
@@ -192,24 +269,160 @@ async def _execute(self, ctx: handler.HandlerContext, requires: dict[ResourceIdS
                 provider.close()
 
+    async def _report_resource_status(
+        self,
+        env_id: uuid.UUID,
+        resource_id: Id,
+        resource_status: const.ResourceState,
+        resource_change: const.Change,
+        ctx: handler.HandlerContext,
+    ) -> tuple[Optional[const.ResourceState], const.Change, dict[ResourceVersionIdStr, dict[str, AttributeStateChange]]]:
+        changes: dict[ResourceVersionIdStr, dict[str, AttributeStateChange]] = {resource_id.resource_version_str(): ctx.changes}
+        response = await self.client.resource_deploy_done(
+            tid=env_id,
+            rvid=resource_id.resource_version_str(),
+            action_id=ctx.action_id,
+            status=resource_status,
+            messages=ctx.logs,
+            changes=changes,
+            change=resource_change,
+        )
+        if response.code != 200:
+            LOGGER.error("Resource status update failed %s", response.result)
+
+        return resource_status, resource_change, changes
+
+    async def execute(
+        self,
+        gid: uuid.UUID,
+        resource_id: Id,
+        resource: dict[str, object],
+        env_id: uuid.UUID,
+        reason: str,
+    ) -> tuple[Optional[const.ResourceState], const.Change, dict[ResourceVersionIdStr, dict[str, AttributeStateChange]]]:
+        started: datetime.datetime = datetime.datetime.now().astimezone()
+        # TODO: double check: is this required? Is failure handled properly?
+
+        try:
+            resource: Resource = Resource.deserialize(resource)
+        except Exception:
+            msg = data.LogLine.log(level=const.LogLevel.ERROR, msg="Unable to deserialize %(resource_id)s", resource_id=resource_id.resource_version_str(), timestamp=datetime.datetime.now().astimezone()),
+
+            await self.client.resource_action_update(
+                tid=env_id,
+                resource_ids=[resource_id.resource_version_str()],
+                action_id=uuid.uuid4(),
+                action=const.ResourceAction.deploy,
+                started=started,
+                finished=datetime.datetime.now().astimezone(),
+                status=const.ResourceState.unavailable,
+                messages=[msg]
+            )
+
+        ctx = handler.HandlerContext(resource)
+        ctx.debug(
+            "Start run for resource %(resource)s because %(reason)s",
+            resource=str(resource_id),
+            deploy_id=gid,
+            agent=self.agent.name,
+            reason=reason,
+        )
+
+        try:
+            requires: dict[ResourceIdStr, const.ResourceState] = await self.send_in_progress(
+                ctx.action_id, env_id, resource_id.resource_version_str()
+            )
+        except Exception:
+            ctx.set_status(const.ResourceState.failed)
+            ctx.exception("Failed to report the start of the deployment to the server")
+        else:
+            await self._execute(resource, ctx=ctx, requires=requires)
```

Comment:
I find it more readable if you add this to the end of the try-except block.

(for me, the main flow of the code should be visually apparent, not hidden behind an exception handling block)

Reply:
I don't see how this would look like (assuming we want to keep the current behavior of only running `self._execute` if no exceptions were raised during `send_in_progress`)

Reply:
brain fart: just added a return in the except block